src/status_im/contexts/keycard/migrate/events.cljs

```diff
@@ -0,0 +1,125 @@
+         {:keys [masked-phrase pin]}            (get-in db [:keycard :migration])]
+     (cond
+
+       (not initialized?)
+       {:fx [[:keycard/init-card
+              {:pin        pin
+               :on-success #(get-application-info-and-continue key-uid)}]]}
+
+       (not has-master-key?)
+       {:fx [[:effects.keycard/generate-and-load-key
+              {:mnemonic   (security/safe-unmask-data masked-phrase)
+               :pin        pin
+               :on-success #(get-application-info-and-continue key-uid)}]]}
+
+       :else
+       {:fx [[:effects.keycard/get-keys
+              {:pin        pin
+               :on-success #(rf/dispatch [:keycard/migration.convert-to-keycard-profile %])
+               :on-failure #()}]]}))))
+
+(rf/reg-event-fx :keycard/migration.start
+ (fn [{:keys [db]}]
+   {:fx [[:dispatch
+          [:keycard/connect
+           {:key-uid (get-in db [:profile/profile :key-uid])
+            :on-success #(rf/dispatch [:keycard/migration.continue])
+            :on-error
+            (fn [error]
+              (if (= error :keycard/error.keycard-blank)
+                (rf/dispatch [:keycard/migration.continue])
+                (rf/dispatch [:keycard/on-application-info-error error])))}]]]}))
+
+(rf/reg-event-fx :keycard/migration.get-phrase
+ (fn [{:keys [db]}]
+   {:db (assoc-in db [:keycard :migration] nil)
+    :fx [[:dispatch [:navigate-back]]
+         (if (string/blank? (get-in db [:profile/profile :mnemonic]))
+           [:dispatch
+            [:open-modal :screen/use-recovery-phrase
+             {:on-success #(rf/dispatch [:keycard/migration.phrase-entered %])}]]
+           [:dispatch
+            [:open-modal :screen/backup-recovery-phrase
+             {:on-success #(rf/dispatch [:keycard/migration.phrase-backed-up %])}]])]}))
+
+(rf/reg-event-fx :keycard/migration.phrase-entered
+ (fn [{:keys [db]} [{:keys [phrase]}]]
+   {:db (assoc-in db [:keycard :migration :masked-phrase] (security/mask-data phrase))
+    :fx [[:dispatch [:navigate-back]]
+         [:dispatch
+          [:open-modal :screen/keycard.authorise
+           {:on-success #(rf/dispatch [:keycard/migration.authorisation-success %])}]]]}))
+
+(rf/reg-event-fx :keycard/migration.phrase-backed-up
+ (fn [{:keys [db]}]
+   {:db (assoc-in db
+         [:keycard :migration :masked-phrase]
+         (security/mask-data (get-in db [:profile/profile :mnemonic])))
+    :fx [[:dispatch [:profile.settings/profile-update :mnemonic nil]]
+         [:dispatch [:navigate-back]]
+         [:dispatch
+          [:open-modal :screen/keycard.authorise
+           {:on-success #(rf/dispatch [:keycard/migration.authorisation-success %])}]]]}))
+
+(rf/reg-event-fx :keycard/migration.authorisation-success
+ (fn [{:keys [db]} [masked-password]]
+   (let [{:keys [initialized?]} (get-in db [:keycard :application-info])]
+     {:db (assoc-in db [:keycard :migration :masked-password] masked-password)
+      :fx [[:dispatch
+            [:show-bottom-sheet
```

Comment:
yes that's correct, but bottom sheet is an exception, because we need to dispatch it from event namespaces, we could use keywords similar to screens, but im not sure it worth it

Reply:
bottom sheet  is more like screen, there is no much hiccup, it's just component, but hiccup used to pass parameters, there are tradeoffs, in most cases we want to use it from view, and we have lots of compact UI to show and if will use keywords for them it's not really convenient, we could create a task and think how we could improve it, but for me current implementation is kinda ok, actually in the status 1.0 we had mixed approach with hiccup and keywords, but it was confusing little bit

Reply:
The standard model `UI = f(states)` solves this type of problem very well, so I don't see much need for thinking of a different approach. The guideline would be sort of _no UI components in events_ (no require namespaces from views, we should have a linter for that). I'm always afraid of broken windows, one example here and there, plus a large team, and certain practices start to spread as if they are okay.

> we could create a task and think how we could improve it

Good idea 👍🏼